oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrantBuilder.java

```diff
@@ -161,45 +158,44 @@ final Duration refreshBefore() {
     }
 
     /**
-     * A {@link Supplier} to load Access Token from, to be able to restore the previous session. OPTIONAL.
-     */
-    @SuppressWarnings("unchecked")
-    public final T tokenPersistencySupplier(Supplier<? extends GrantedOAuth2AccessToken> tokenSupplier) {
-        this.tokenSupplier = requireNonNull(tokenSupplier, "tokenSupplier");
-        return (T) this;
-    }
-
-    @Nullable
-    final Supplier<? extends GrantedOAuth2AccessToken> tokenPersistencySupplier() {
-        return tokenSupplier;
-    }
-
-    /**
-     * A {@link Consumer} to store Access Token to, to be able restore the previous session. OPTIONAL.
+     * An optional {@link Supplier} to acquire an access token before requesting it to the authorization server.
+     * If the provided {@link GrantedOAuth2AccessToken} is valid, the client doesn't request a new token.
+     *
+     * <p>This is supposed to be used with {@link #newTokenConsumer(Consumer)} and gets executed
+     * in the following cases:
+     * <ul>
+     *     <li>Before the first attempt to acquire an access token.
+     *     <li>Before a subsequent attempt after token issue or refresh failure.
+     * </ul>
+     * @see #newTokenConsumer(Consumer)
      */
     @SuppressWarnings("unchecked")
-    public final T tokenPersistencyConsumer(Consumer<? super GrantedOAuth2AccessToken> tokenConsumer) {
-        this.tokenConsumer = requireNonNull(tokenConsumer, "tokenConsumer");
+    public final T fallbackTokenProvider(
+            Supplier<CompletableFuture<? extends GrantedOAuth2AccessToken>> fallbackTokenProvider) {
+        this.fallbackTokenProvider = requireNonNull(fallbackTokenProvider, "fallbackTokenProvider");
         return (T) this;
     }
 
     @Nullable
-    final Consumer<? super GrantedOAuth2AccessToken> tokenPersistencyConsumer() {
-        return tokenConsumer;
+    public final Supplier<CompletableFuture<? extends GrantedOAuth2AccessToken>> fallbackTokenProvider() {
+        return fallbackTokenProvider;
     }
 
     /**
-     * An optional {@link Executor} that facilitates asynchronous access token obtain and refresh operations.
+     * An optional hook which gets executed whenever a new token is issued.
+     *
+     * <p>This can be used in combination with {@link #fallbackTokenProvider(Supplier)} to store a newly issued
+     * access token which will then be retrieved by invoking {@link #fallbackTokenProvider}.
      */
     @SuppressWarnings("unchecked")
-    public final T executor(Executor executor) {
-        this.executor = requireNonNull(executor, "executor");
+    public final T newTokenConsumer(Consumer<? super GrantedOAuth2AccessToken> newTokenConsumer) {
+        this.newTokenConsumer = requireNonNull(newTokenConsumer, "newTokenConsumer");
         return (T) this;
     }
 
     @Nullable
-    final Executor executor() {
-        return executor;
+    public final Consumer<? super GrantedOAuth2AccessToken> newTokenConsumer() {
```

Comment:
Could we hide this from the public API for consistency? Our builders don't have getters.

Reply:
Thanks for the correction, updated!